tests/unit_tests/training/mimo/test_mimo_config.py

```
@@ -0,0 +1,206 @@
+
+def test_module_parallelism_finalize_computes_dp():
+    parallelism = ModuleParallelismConfig(tensor_model_parallel_size=2, pipeline_model_parallel_size=2)
+    parallelism.finalize(world_size=16)
+    assert parallelism.data_parallel_size == 4
+    assert parallelism.total_model_parallel_size == 4
+    assert parallelism.total_ranks == 16
+
+
+def test_module_parallelism_finalize_invalid_world_size():
+    parallelism = ModuleParallelismConfig(tensor_model_parallel_size=3, pipeline_model_parallel_size=2)
+    with pytest.raises(ValueError, match="world_size .* not divisible"):
```

Ruff (0.15.0)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)